File: spring_security/eg05_sb3_oauth2_resource_server_jwt_symmetric_key/src/main/java/com/sc/sb/auth/service/ScTokenGeneratorService.java
package com.sc.sb.auth.service;

import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.jose.jws.MacAlgorithm;
import org.springframework.security.oauth2.jwt.JwsHeader;
import org.springframework.security.oauth2.jwt.JwtClaimsSet;
import org.springframework.security.oauth2.jwt.JwtEncoder;
import org.springframework.security.oauth2.jwt.JwtEncoderParameters;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ScTokenGeneratorService {

    private final JwtEncoder encoder;

    public ScTokenGeneratorService(JwtEncoder encoder) {
        this.encoder = encoder;
    }

    public String generateToken(Authentication authentication, String[] requestedScopes) {
        Instant now = Instant.now();

        List<String> requestedScopeList = List.of(requestedScopes);

        String scope = authentication.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .filter(requestedScopeList::contains) // Add only the requested scopes
                .collect(Collectors.joining(" "));

        JwtClaimsSet claims = JwtClaimsSet.builder()
                .issuer("self")
                .issuedAt(now)
                .expiresAt(now.plus(1, ChronoUnit.HOURS))
                .subject(authentication.getName())
                .claim("scope", scope)
                .build();

        var encoderParameters = JwtEncoderParameters.from(JwsHeader.with(MacAlgorithm.HS512).build(), claims);

        return this.encoder.encode(encoderParameters).getTokenValue();
    }


}

File: spring_security/eg06_kotlin_sb3_oauth2_resource_server_jwt_symmetric_key/src/main/kotlin/com/sc/sb/auth/ScTokenGeneratorService.kt
package com.sc.sb.auth

import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.jose.jws.MacAlgorithm
import org.springframework.security.oauth2.jwt.JwsHeader
import org.springframework.security.oauth2.jwt.JwtClaimsSet
import org.springframework.security.oauth2.jwt.JwtEncoder
import org.springframework.security.oauth2.jwt.JwtEncoderParameters
import org.springframework.stereotype.Service
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.stream.Collectors

@Service
class ScTokenGeneratorService(private val encoder: JwtEncoder) {

    fun generateToken(authentication: Authentication, requestedScopes: Array<String>): String {
        val now = Instant.now()

        val scope = authentication.authorities.stream()
            .map { it.authority }
            .filter { requestedScopes.contains(it) }
            .collect(Collectors.joining(" "))

        val claims = JwtClaimsSet.builder()
            .issuer("self")
            .issuedAt(now)
            .expiresAt(now.plus(1, ChronoUnit.HOURS))
            .subject(authentication.name)
            .claim("scope", scope)
            .build()

        val encoderParameters = JwtEncoderParameters.from(JwsHeader.with(MacAlgorithm.HS512).build(), claims)

        return this.encoder.encode(encoderParameters).tokenValue
    }
}